BaseZapPlugin and KeyValueGetter

tachibanayui opened the pull request:
Add BaseZapPlugin containing some frequently used methods. 
Add KeyValueGetter (EnvironmentGetter is one of the implementations)
Move Helper classes from other repos that can be applied anywhere

VegetalDev committed:
added some features to support collision-patch on arena-api repository: new methods added to Vectors; ported Direction from arena-api

VegetalDev committed:
Update README.md

VegetalDev committed:
Merge pull request #1 from zapzombies/patch

Stuff to support latest update on Arena-Api

tachibanayui committed:
Implementing Key-value getter

tachibanayui committed:
Base zap plugin

tachibanayui committed:
Add ReflectionUtils javadocs

tachibanayui committed:
Fix getSuperclassTypeParams algorithm

tachibanayui committed:
Leech from various repos

tachibanayui committed:
Small fixes

tachibanayui committed:
improvements

VegetalDev commented on src/main/java/io/github/zap/commons/BaseZapPlugin.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * A base plugin for *all* zap plugin projects
+ */
+public abstract class BaseZapPlugin extends JavaPlugin {
```
IMO this should also have overrides for onDisable() and onLoad(), along with more generalized code for timing so we could reuse the same code for enable, disable, load

tachibanayui replied:
Currently idk what to add about `onDisable` and `onLoad` maybe another timer?

VegetalDev replied:
Yeah I think they just should have a timer. You can probably make a private method to avoid duplicating code

VegetalDev commented on src/main/java/io/github/zap/commons/MultiBoundingBox.java:
```diff
@@ -0,0 +1,203 @@
+ * Encapsulates a composite bounded area made up of any number of axis-aligned bounding boxes, which are not required
+ * to be continuous and may also overlap.
+ */
+public class MultiBoundingBox implements Iterable<BoundingBox> {
```
oh cool you also added this class 👍

tachibanayui replied:
xd

VegetalDev commented on src/main/java/io/github/zap/commons/keyvaluegetters/AbstractKeyValueGetter.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Override
+    public void bind(Object o) throws GetFailedException {
```
i have to say, i don't really like that the usage of var which is inconsistent with the rest of our codebase (that doesn't use var)

tachibanayui replied:
lol I started to use it everywhere. Fixing them kinda takes some time, should we change all to var or to explicit type?

VegetalDev replied:
it is easier IMO to change a relatively small # of usages of var than our entire project of 40000+ lines across repos

tachibanayui replied:
well, we are rewriting a lot of stuff later, so this is an opportunity to use the new language feature?

VegetalDev replied:
we are rewriting a lot of stuff, yes, but there's also a lot of stuff that isn't going to be totally rewritten (my AI lib, for example, which is like 3-4k lines on its own i think); so there's lots of stuff that we would need to go back to and add var for consistency

also IMO var hurts readability a lot; for example in this line

`var rawValue = getResult.getRight().result().get();`

i have no idea what rawValue actually is (i would need to go back and look at how getResult is declared, which isn't the hardest thing in the world, but it adds an unnecessary layer of obfuscation)

of course i believe intelliJ would show you what the type actually is, but when i'm reviewing a PR i typically just read the source in Github. also i would say that relying on IDE features for readability is inherently not a good thing

the only real argument for var is using it in the odd instance where you have an extremely long or complex generic type, but those are rare and i've never really seen an instance where i thought var added readability rather than taking it away

tachibanayui replied:
ok i gonna revert it to explicit type

VegetalDev commented on src/main/java/io/github/zap/commons/keyvaluegetters/AbstractKeyValueGetter.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Override
+    public <T> T get(Class<T> type) throws GetFailedException {
```
i think `type` should be annotated with @NotNull and this goes for any publicly accessible parameter that will throw a NullPointerException/result in unexpected behavior if null is passed

VegetalDev commented on src/main/java/io/github/zap/commons/keyvaluegetters/AbstractKeyValueGetter.java:
```diff
@@ -0,0 +1,146 @@
+            faultyDeserializeOperation.sort((o1, o2) -> prioritizeRequiredSort(o1.getLeft(), o2.getLeft()));
+            faultyBindOperation.sort((o1, o2) -> prioritizeRequiredSort(o1.getLeft(), o2.getLeft()));
+
+            StringBuilder sb = new StringBuilder();
```
sb :kindacringeiflikousesit:

tachibanayui replied:
![image](https://user-images.githubusercontent.com/33594017/134627740-d0841630-e06a-4251-9080-d65f47699cad.png)

VegetalDev commented on src/main/java/io/github/zap/commons/keyvaluegetters/ValueDeserializer.java:
```diff
@@ -0,0 +1,7 @@
+import java.lang.reflect.ParameterizedType;
+
+public interface ValueDeserializer<T> {
+    OperationResult<T> deserialize(String value, ParameterizedType pt, DeserializerEngine deserializers);
```
I suspect that OperationResult<T> will  never normally be null so this should be @NotNull

tachibanayui replied:
yep. I just realized I forget to add a lot of `@Nullable` and `@NotNull`

VegetalDev commented on src/test/java/io/github/zap/commons/utils/ReflectionUtilsTest.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class ReflectionUtilsTest {
```
since this is the only test you've written, have you made sure the other non-trivial stuff works?

tachibanayui replied:
No, I made the test solely for the Reflection stuff because it is so damn hard to debug recursion. I will eventually add other tests but for now, my time is kinda run out

tachibanayui committed:
Add `@NotNull`, `@Nullable` and checks
Added TimeUtils.measure
Add additional timer method in BaseZapPlugin.java
Convert var into explicit type
